### PR TITLE
Fix race conditions in LOG_EVERY_N

### PR DESCRIPTION
**Pull request by aesophor**
Fixes issue #439 ~~by protecting critical sections with mutexes.~~

Note:
1. Use CMakeLists.txt `CheckCXXSourceCompiles` to check the availability of `std::atomic<>`
2. The atomic operations will be handled differently based on the following situations:
    * Compiler supports C++11 - simply declare the integer counters as `std::atomic<int>`
    * Pre C++11 (MSVC) - InterlockedIncrement(), InterlockedExchangeSubtract()
    * Pre C++11 (GCC) - __sync_add_and_fetch(), __sync_sub_and_fetch()

### CODE REVIEW

**Comment by sergiud** [comment]
A mutex is quite heavy. Would thread-local storage possibly suffice?

cc @os12

**Comment by os12** [comment]
There is no need for a mutex. These counters just need to become `std::atomic` (or equivalent non-portable).

👎

**Comment by sergiud** [comment]
Glog still uses C++03. There should be atleast a fallback mechanism for pre C++11 compilers.

File: src/glog/logging.h.in [hunk 1]
```diff
@@ -897,7 +897,7 @@ PLOG_IF(FATAL, GOOGLE_PREDICT_BRANCH_NOT_TAKEN((invocation) == -1))    \
 #define LOG_OCCURRENCES_MOD_N LOG_EVERY_N_VARNAME(occurrences_mod_n_, __LINE__)
 
 #define SOME_KIND_OF_LOG_EVERY_N(severity, n, what_to_do) \
-  static int LOG_OCCURRENCES = 0, LOG_OCCURRENCES_MOD_N = 0; \
+  thread_local int LOG_OCCURRENCES = 0, LOG_OCCURRENCES_MOD_N = 0; \
```
Comment (os12) [inline src/glog/logging.h.in]: This looks wrong to me. Imagine the same `LOG_EVERY_XXX()` called from several threads concurrently - every thread will get its own count now...

Comment (aesophor) [reply src/glog/logging.h.in]: I've made the counters `std::atomic<int>`, and now the only thing left is a fallback mechanism for pre C++11 compilers. Do you have any idea?

Comment (sergiud) [reply src/glog/logging.h.in]: Both GCC and MSVC provide atomic intrinsics. Please see https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html and https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedincrement.

File: src/glog/logging.h.in [hunk 2]
```diff
@@ -86,6 +86,22 @@
...
+// Check if it's compiled in C++11 mode.
+//
+// GXX_EXPERIMENTAL_CXX0X is defined by gcc and clang up to at least
+// gcc-4.7 and clang-3.1 (2011-12-13).  __cplusplus was defined to 1
+// in gcc before 4.7 (Crosstool 16) and clang before 3.1, but is
+// defined according to the language version in effect thereafter.
+// Microsoft Visual Studio 14 (2015) sets __cplusplus==199711 despite
+// reasonably good C++11 support, so we set LANG_CXX for it and
+// newer versions (_MSC_VER >= 1900).
+#define SUPPORT_CPLUSPLUS11 (defined(__GXX_EXPERIMENTAL_CXX0X__) || __cplusplus >= 201103L || \
```
Comment (sergiud) [inline src/glog/logging.h.in]: `std::atomic<>` availability should be checked in `CMakeLists.txt` using `CheckCXXSymbolExists` or `CheckCXXSourceCompiles` module.

Comment (aesophor) [reply src/glog/logging.h.in]: Will do!

File: src/glog/logging.h.in [hunk 3]
```diff
@@ -924,14 +931,99 @@ PLOG_IF(FATAL, GOOGLE_PREDICT_BRANCH_NOT_TAKEN((invocation) == -1))    \
...
+        __FILE__, __LINE__, @ac_google_namespace@::GLOG_ ## severity, LOG_OCCURRENCES, \
+        &what_to_do).stream()
+
+#define SOME_KIND_OF_LOG_IF_EVERY_N(severity, condition, n, what_to_do) \
+  static int LOG_OCCURRENCES = 0, LOG_OCCURRENCES_MOD_N = 0; \
+  __sync_add_and_fetch(&LOG_OCCURRENCES, 1); \
+  if (condition && \
+      (__sync_add_and_fetch(&LOG_OCCURRENCES_MOD_N, 1) || true) && \
+      ((LOG_OCCURRENCES_MOD_N >= n && __sync_sub_and_fetch(&LOG_OCCURRENCES_MOD_N, n)) || true) && \
+      LOG_OCCURRENCES_MOD_N == (1 % n)) \
```
Comment (aesophor) [inline src/glog/logging.h.in]: This if statement was originally as follows:
```
if (condition && \
      ((LOG_OCCURRENCES_MOD_N=(LOG_OCCURRENCES_MOD_N + 1) % n) == (1 % n))) \
```

The rewritten version consists of 4 conditions (connected by logical AND).
1. does the argument  `condition` evaluate to true?
2. perform atomic `++LOG_OCCURRENCES_MOD_N`
3. if `LOG_OCCURRENCES_MOD_N >= n`, then perform atomic `LOG_OCCURRENCES_MOD_N -= n`
4. finally, is `LOG_OCCURRENCES_MOD_N == (1 % n)` ?

**Comment by sergiud** [comment]
@aesophor Could you please resolve the conflict?

**Comment by aesophor** [comment]
@sergiud I've resolved the conflict and CI has passed successfully. Thanks!

**Pull request merged**